src/ProducerEventProcessor/UploaderCaptureEventCollector.cpp

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "ProducerEventProcessor/UploaderCaptureEventCollector.h"
+
+#include "Introspection/Introspection.h"
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/MakeUniqueForOverwrite.h"
+#include "OrbitBase/ThreadUtils.h"
+#include "capture.pb.h"
+#include "services.grpc.pb.h"
+
+using orbit_capture_uploader::DataReadiness;
+using orbit_grpc_protos::CaptureResponse;
+using orbit_grpc_protos::ClientCaptureEvent;
+
+namespace orbit_producer_event_processor {
+
+UploaderCaptureEventCollector::UploaderCaptureEventCollector() {
+  // Create an output stream which will be used to convert capture events into parts of a well
+  // formatted capture file. This should never fail.
+  capture_data_buffer_stream_ = std::make_unique<orbit_capture_file::BufferOutputStream>();
+  output_stream_ =
+      orbit_capture_file::CaptureFileOutputStream::Create(capture_data_buffer_stream_.get());
+}
+
+UploaderCaptureEventCollector::~UploaderCaptureEventCollector() {
+  LOG("Total number of events uploaded: %u", total_uploaded_event_count_);
+  LOG("Total number of bytes uploaded: %u", total_uploaded_data_bytes_);
+  LOG("Total number of write event errors: %u", total_write_error_count_);
+
+  if (total_uploaded_event_count_ > 0) {
+    float average_bytes = static_cast<float>(total_uploaded_data_bytes_) /
+                          static_cast<float>(total_uploaded_event_count_);
+    LOG("Average number of bytes per event: %.2f", average_bytes);
+  }
+}
+
+void UploaderCaptureEventCollector::Stop() {
+  absl::MutexLock lock{&mutex_};
+
+  // Protect stop_requested_ with mutex_ so that we can use stop_requested_ in Conditions for
+  // Await/LockWhen.
+  stop_requested_ = true;
+
+  // Capture finish event would be processed before requesting stopping this collector, and the
+  // output stream should already be closed while processing the capture finish event.
+  CHECK(!output_stream_->IsOpen());
+}
+
+void UploaderCaptureEventCollector::AddEvent(ClientCaptureEvent&& event) {
+  absl::MutexLock lock{&mutex_};
+
+  // Drop events received after "stop capture" being requested.
+  if (stop_requested_) return;
+
+  // The output stream will be closed while processing the capture finish event. Drop events
+  // received after closing the output stream.
+  CHECK(output_stream_ != nullptr);
+  if (!output_stream_->IsOpen()) return;
+
+  auto write_result = output_stream_->WriteCaptureEvent(event);
+  if (write_result.has_error()) {
+    total_write_error_count_++;
+    return;
+  }
```

Comment:
Hmm. When does this happen? Shouldn't we abort the whole thing on a write error?

Reply:
Yes, we are aborting things if there is a write failure. I just use this to collect some statistics.